### ISSUE TITLE
Add `requirements.txt`? and other questions & suggestions

### ISSUE THREAD

**Issue by phseiff** [theendlessriver13/gh-md-to-moodle#10]
Hello,
I'm the maintainer of `gh-md-to-html` and I have some questions and suggestions regarding `gh-md-to-moodle`, if that's okay with you :)

I noticed `gh-md-to-moodle` declares its dependencies [in your `setup.cfg`-file](https://github.com/theendlessriver13/gh-md-to-moodle/blob/master/setup.cfg#L24-L28). Whilst this is (as far as I'm aware?) the recommended method, it is neither recognized by dependabot, nor is it recognized by the GitHub Dependency Graph, who both only look at `setup.py` and `requirements.txt`.
Having the project's dependencies recognized by the GitHub Dependency Graph means that (a) you are notified of potential vulnerabilities in your dependencies if they arise, (b) your dependencies are publically and easily visible in the Insights-Tab of your repository, and (c) your project gets listed in the dependents of the repositories you depend on, which potentially helps their maintainers (like me) become aware of your usage of your tool (potentially useful if they want to learn more about their user base), plus it looks good in their repositories (which actually is the main reason I decided to file this issue for).
You might consider additionally adding your dependencies to a `requirements.txt`-file for the reasons above.

Another thing I noticed is that you [strip the newline after and in front of opening and closing `code`-tags](https://github.com/theendlessriver13/gh-md-to-moodle/blob/master/gh_md_to_moodle/main.py#L20-L23). I'm not exactly sure why you are doing this, but if this serves to compensate for a bug in `gh-md-to-html`, then I'd be happy to fix said bug! :D

It also seems like you inject some JavaScript into the generated page, that then embeds some CSS into the page's header, which causes Moodle not to display the file correctly in the preview (according to your README). Did you consider using the `extra_css` argument in `gh_md_to_html.main()` to add some additional HTML like `gh-md-to-html`'s README suggests? It might work too (unless you need the CSS to be specifically part of the header of the page), and display better in the preview.

It also seems to me like you are [trying to reverse](https://github.com/theendlessriver13/gh-md-to-moodle/blob/master/gh_md_to_moodle/main.py#L84-L85) the effects of `gh-md-to-html` caching images and changing links to images within the rendered HTML. You can do this supplying `enable_image_downloading=False` to `gh_md_to_html.main`, which is probably a cleaner solution (not properly documented yet, though, admittedly).

I'm also somewhat confused about the nature of the CSS you inject: Is it merely supposed to make the generated page blend in with Moodle better (as seems to be the case with the pink color for code blocks), or does it also fix potential problems with the styling `gh-md-to-html` injects? If it's (partially) the later one, feel free to file an issue in `gh-md-to-html` (or tell me about it in this issue) so I can try to fix them. :)

**Comment by theendlessriver13** [comment]
Hi, thanks for your comments and for the great tool you made I really enjoy using it!


sure, I can add a `requirements.txt` in the next days, or you can just send a PR if you want to. Github unfortunatelly refuses to implement proper parsing of `setup.cfg` this has been a known issue for quite some time now...

>Another thing I noticed is that you strip the newline after and in front of opening and closing code-tags. 

I think there was an issue with the code formatter I used from `BeatufilSoup`. It would create newlines after the code tags which would cause the page not rendering correctly -- not sure why this is the case though. I did not want to put too much effort into this so I must admit this is a hacky solution.


this would be the formatted `html` output:
```html
<p>
Assign one of the following data types
<code>
int
</code>
,
<code>
float
</code>
,
<code>
str
</code>
,
<code>
bool
</code>
to the statements below:
</p>
```
but it would render as:
![grafik](https://user-images.githubusercontent.com/54631600/114326915-4189f980-9b37-11eb-921d-107bf21d7ee1.png)

> It also seems like you inject some JavaScript into the generated page

This whole solution is trying to work with the little options moodle provides for easily providing syntax highlighted code snippets. So the issue I initially wrote this tool for was me having to create exercises for programming courses. So it is really convenient to write markdown for this, but moodle can't handle it natively. Of course I could have provided pdfs, but I wanted to integrate it so no files had to be downlaoded. Moodle has a built in text editor where you can edit the page you are creating either in formatted text where the html is done behind the scenes or as plain html. This editor/the whole page does not allow additional style sheets (they are removed when parsing your provided input). But moodle allows `<script>` tags in there. I  think they call it a feature for integrating fancy js based tools, but you can also call it an XSS vulnerability (which I am happy exists 😄)... The reason the page does not show up in the editor's preview fully rendered (as I said in the readme) is, that the `js` is not executed. It is only executed on page load when the finished and deployed page is loaded by the users.

> It also seems to me like you are trying to reverse the effects of gh-md-to-html caching images and changing links to images within the rendered HTML. You can do this supplying enable_image_downloading=False to gh_md_to_html.main, which is probably a cleaner solution (not properly documented yet, though, admittedly).

you are probably right, I will try this and fix it in the future. I did not know about the option, I must admit I got a little overwhelmed by all the options :).

> I'm also somewhat confused about the nature of the CSS you inject

That's what I was also partitially explaining above.
Moodle itself uses `bootstrap` iirc but the code tags are also defined by the github css iirc. I liked the bootstrap pink, so just personal preference here. Everything else is modifying it so it renders properly inside of moodle.

I know I am not using the supported interface of `gh-md-to-html` here so I don't expect it to not brek at anytime :). I must also say I am extremely grateful for you writing and providing your tool, this helped me a lot. So my thingy is really just a thin wraper around `gh-md-to-html` to make it work (more easily) for my edge case so I am probably the only one using it to (ab)use moodle . 

If you want me to provide more credit or do it this differently, I am happy to do this and open for suggestions.

**Comment by phseiff** [comment]
> I really enjoy using it!

Thanks, I'm glad to hear that!

> sure, I can add a requirements.txt in the next days, or you can just send a PR if you want to.

Thanks, that's great! :D

> sure, I can add a `requirements.txt` in the next days, or you can just send a PR if you want to. Github unfortunatelly refuses to implement proper parsing of `setup.cfg` this has been a known issue for quite some time now...

Yeah, I read about this somewhere.... they didn't even manage to get it into dependabot :/

> I think there was an issue with the code formatter I used from `BeatufilSoup`. It would create newlines after the code tags which would cause the page not rendering correctly -- not sure why this is the case though. I did not want to put too much effort into this so I must admit this is a hacky solution.

Ooooh okay, I see. That's good to know; I will keep this in mind in case I ever implement a prettify-option in gh-md-to-html.

> This whole solution is trying to work with the little options moodle provides for easily providing syntax highlighted code snippets [...]

Oh, so that's how it works. Thanks for explaining and contextualizing it to me!

> I did not know about the option

It's completely undocumented as of now, so that's on me :sweat_smile:

> I must admit I got a little overwhelmed by all the options :).

Thank you for the valuable feedback; I will improve the usage-section of the README in the future, then! :)

> I know I am not using the supported interface of `gh-md-to-html` here so I don't expect it to not brek at anytime :). I must also say I am extremely grateful for you writing and providing your tool, this helped me a lot. So my thingy is really just a thin wraper around `gh-md-to-html` to make it work (more easily) for my edge case so I am probably the only one using it to (ab)use moodle .

It doesn't look too unstable to me, and to be honest, `gh-md-to-html` also abuses some behaviors of GitHub's underlying REST API, so you're in good company here :laughing: 

> If you want me to provide more credit or do it this differently, I am happy to do this and open for suggestions.

Don't worry, it's cool as it is (after all, you even have a link to my project in your README).

---

Thanks for using my module and your patience in answering my questions, then, and thanks for your feedback, input & contextualization! :)

**Comment by theendlessriver13** [comment]
I am just testing the `enable_image_downloading` it looks like the image is downloaded and then uploaded to gihub or am I mistaken?  I'd rather prefer to control myself where things are uploaded, is there also an option for this?
```html
<img alt="alt_text" data-canonical-src="https://i.fluffy.cc/KT532NcD7QsGgd6zTXsbQsm6lX7sMrCD.png" src="https://camo.githubusercontent.com/95ca892d25ea550ffd2dc788a33f8c3465247e546ea6d399e642ddd1b5965dcb/68747470733a2f2f692e666c756666792e63632f4b543533324e6344375173476764367a5458736251736d366c5837734d7243442e706e67" style="max-width:100%;"/>
```
This test is failing because I am expecting both urls to be the same: https://github.com/theendlessriver13/gh-md-to-moodle/blob/master/tests/main_test.py#L157-L176
It would also be fine to just have `src`. Is `data-canoical-src` a special gh thing?

**Comment by phseiff** [comment]
> it looks like the image is downloaded and then uploaded to gihub or am I mistaken?

It isn't downloaded and uploaded (only the markdown is uploaded, never images or anything like that).
What happens is, I think, that GitHub realizes that the image is hosted online (by trying to access its `src`), and then caches it, stores the original `src` as the `data-canonical-src`, and uses the URL under which GitHub hosts the cached image as the new `src`.

This means that yes, GitHub caches images, but only those that are hosted online and accessible by GitHub anyways. If you use a relative link to an image, for example `![alt](test.png)`, the resulting image tag is simply `<img src="test.png">` since GitHub can't cache images from your local machine.

> It would also be fine to just have `src`. Is `data-canoical-src` a special gh thing?

Yes, I think so. It's where GitHub stores the link to the original image before they cached it (you can also see it if you look at the page source of a repository front page of a repo that contains images in its README), as explained above.

---

I fixed this in `v1.9.3` right now, so you should be able to update to that version now, and `src` should always point to what the markdown image link pointed to, without the special caching-link GitHub replaces it with.
GitHub will still cache images that are hosted online for an unspecified amount of time (that's not avoidable with the GitHub REST API, apparently), but this shouldn't be an issue when the file is hosted open in the web anyways.

**Comment by theendlessriver13** [comment]
`1.9.2` and `1.9.3` are broken for me... 
I cannot import your function as I did before:
```console
   from gh_md_to_html import main as convert_md
ImportError: cannot import name 'main' from 'gh_md_to_html' (unknown location)
```
also: using `requirements.txt` in [`setup.py`](https://github.com/phseiff/github-flavored-markdown-to-html/blob/master/setup.py#L19) makes it impossible to install your package from the pypi tarball, since you did not include `requirements.txt` in your `package_data` or have a `MANIFEST.in`. I'd suggest to just use the [`install_requires`](https://python-packaging.readthedocs.io/en/latest/dependencies.html#specifying-dependencies)

It looks like you stopped using tags in your repo which makes it hard to compare different versions. It took me a while to figure out what broke it and I think it was you removing the `__init__.py`.
```diff
diff --git a/src/__init__.py b/src/__init__.py
new file mode 100644
index 0000000..a86c6bb
--- /dev/null
+++ b/src/__init__.py
@@ -0,0 +1 @@
+from .__main__ import *
```
So it would be nice if you could add it back ;-)
After adding it back, it works fine.

I noticed something else: links are now escaped, which is a problem. They must remain like this: `https://foo-bar.baz?test=foo&bar=wat` so `&` is not excaped.
```html
<a href="https://foo-bar.baz?test=foo&amp;bar=wat" rel="nofollow" target="_blank">
<img alt="alt_text" data-canonical-src="https://foo-bar.baz?test=foo&amp;bar=wat" src="https://foo-bar.baz?test=foo&amp;bar=wat" style="max-width:100%;"/>
```
When you fix this, I can remove all my image url replacement shenaningans -- thank you 👍

**Comment by phseiff** [comment]
> also: using `requirements.txt` in [`setup.py`](https://github.com/phseiff/github-flavored-markdown-to-html/blob/master/setup.py#L19) makes it impossible to install your package from the pypi tarball, since you did not include `requirements.txt` in your `package_data` or have a `MANIFEST.in`.

I added the requirements directly to the `setup.py` now.
Thanks for pointing this out!

> It looks like you stopped using tags in your repo which makes it hard to compare different versions.

I actually never used tags the way they are supposed to. I had a pipeline on GitHub actions that creates a new release on GitHub on every push event, no matter the version number in `setup.py`, and the version number of these releases did not correspond to the ones in `setup.py` at all. In addition to that, I had a second pipeline that was triggered every time a new release on GitHub was released, and attempted to publish said release to PyPi, which either failed (because the version number wasn't incremented) or didn't fail. So the whole thing about creating releases and tags was a completely unnecessary intermediate step amidst a hacky pipeline that added nothing but clutter to the repository, which is why I removed it (the tag versioning didn't even correspond to PyPi versions in the first place, after all).

I will try to integrate GitHub tags properly into my release pipeline once I find time to do so :)

> It took me a while to figure out what broke it and I think it was you removing the `__init__.py`.

Yes, definitely. I accidentally removed it from git when I cleared up some issues with my `.gitignore`.

> I noticed something else: links are now escaped, which is a problem.

I installed an older version (`v1.7.4`) and was able to confirm that this was already the case back then, if my tests reproduce it correctly, so this isn't really a "new" issue.
I also just did some quick googling, and according to stack overflow, html-escaping ampersands in href-attributes [is actually necessary](https://stackoverflow.com/a/3705601). My tests with bs4 confirm that bs4 seems to not even allow to unescape href-attributes (or any other type of attributes, for that matter). It's also worth noting that the `href`-values in question aren't url-escaped, but rather, html-escaped, so this is the intended behavior and doesn't impact a browser's ability to resolve them (I didn't know this too before doing some googling right now, though).

This is what looking at a page source generated by gh-md-to-html looks like when the ampersand in an `href` is unescaped; it's marked red by the browser because it technically isn't valid HTML:
> ![image](https://user-images.githubusercontent.com/31518703/114463159-5c667780-9be4-11eb-85fc-c28870aae190.png)

and even your own tool (gh-md-to-moodle) generates links like this:

>  ![image](https://user-images.githubusercontent.com/31518703/114466653-0f38d480-9be9-11eb-9fef-5f44dacfda49.png)

So yeah, the thing about html-escaping href-attributes seems to be intended (it's caused by bs4, though, so even I was not aware of it until right now).

---

I now created version `1.9.6`, which can be installed just fine (I actually checked by uninstalling and reinstalling this time :sweat_smile: ), including from the `.tar.gz`, though the thing about href-escaping isn't fixed in it since it turned out to be intended (as described above).
